### PR TITLE
feat(automl): redesign confusion matrix and multi-select backtest metrics

### PR DESCRIPTION
**Pull request by jefho-rh**
https://redhat.atlassian.net/browse/RHOAIENG-78193

## Description

Redesigns the AutoML confusion matrix tab and adds multi-select metric support to the backtest window chart.

**Confusion matrix redesign:**
- Replace intensity-based cell coloring (diagonal/off-diagonal with low/medium/high gradients) with a simpler correct-prediction highlight
- Add precision per column header and recall per row header as sub-labels
- Add overall accuracy stat box above the matrix
- Add "PREDICTED CLASS" and "ACTUAL CLASS" axis labels with vertical text for the actual class
- Replace gradient legend bar with a single "Correct prediction" legend dot

https://github.com/user-attachments/assets/9a265c9c-876c-41e7-a4b7-b6546082a30e

**Backtest window chart multi-select:**
- Convert single-select metric dropdown to multi-select with checkboxes
- Add "Show all" option to select/deselect all metrics at once
- Single metric selected: full-width chart (existing layout)
- Multiple metrics selected: 2-column grid of smaller charts, each with its own title and description
- Hide dropdown in print mode, show all selected metrics

https://github.com/user-attachments/assets/1022117d-0422-4628-ad1b-204dc7ecfbab

## How Has This Been Tested?

- Unit tests updated and passing (25 ConfusionMatrixTab tests, 7 BacktestingTab tests)
- Lint and type-check pass across all packages

## Test Impact

- Updated existing confusion matrix tests to match redesigned UI (precision/recall sub-labels, stat box, legend)
- Added test for One vs Rest recall/precision sub-label values
- Added test for `formatPct` zero-total edge case (`0.0%` when row total is zero)

## Request review criteria:

Self checklist (all need to be checked):
- [ ] The developer has manually tested the changes and verified that the changes work
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)
- [x] The code follows our [Best Practices](/docs/best-practices.md) (React coding standards, PatternFly usage, performance considerations)

If you have UI changes:
- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

## Summary by CodeRabbit

* **New Features**
  * Backtesting charts now support multi-metric selection (with “show all”) and can render a single large chart or a grid of per-metric charts.
  * Selected backtest metrics persist across modal tab changes and are included in the print view.
  * Confusion matrices now emphasize “correct prediction” with updated precision/recall and overall accuracy presentation and a clearer legend.
* **Bug Fixes**
  * Improved backtest tooltip positioning for non-default chart widths.
  * Enhanced print layout to reduce unwanted page breaks and added dedicated backtest window card styling.

### CODE REVIEW

**Comment by jefho-rh** [comment]
@coderabbitai review

File: packages/automl/frontend/src/app/components/run-results/AutomlModelDetailsModal/tabs/__tests__/BacktestingTab.spec.tsx [hunk 1]
```diff
@@ -10,7 +10,7 @@ jest.mock(
   '~/app/components/run-results/AutomlModelDetailsModal/components/BacktestWindowChart',
   () => ({
     __esModule: true,
-    default: ({ evalMetric }: { evalMetric: string }) => (
+    default: ({ evalMetric }: { evalMetric: string; print?: boolean }) => (
       <div data-testid="backtest-window-chart">{evalMetric}</div>
     ),
```
[the remark this answers is not in the document]
Comment (jefho-rh) [reply packages/automl/frontend/src/app/components/run-results/AutomlModelDetailsModal/tabs/__tests__/BacktestingTab.spec.tsx:15]: The mock intentionally renders only `evalMetric` to verify data flows through `BacktestingTab`. testing prop-forwarding through a mock is circular. The existing tests already cover that `BacktestWindowChart` receives the right data. No change needed.

File: packages/automl/frontend/src/app/components/run-results/AutomlModelDetailsModal/tabs/ConfusionMatrixTab.tsx [hunk 2]
```diff
@@ -172,146 +162,151 @@ const ConfusionMatrixTab: React.FC<TabContentProps> = ({
...
   const rowTotals = labels.map((row) => labels.reduce((sum, col) => sum + getCell(row, col), 0));
-
-  // Compute per-column totals
   const colTotals = labels.map((col) => labels.reduce((sum, row) => sum + getCell(row, col), 0));
 
-  // Overall accuracy
   const totalCorrect = labels.reduce((sum, label) => sum + getCell(label, label), 0);
   const grandTotal = rowTotals.reduce((sum, t) => sum + t, 0);
-  const overallPct = grandTotal > 0 ? ((totalCorrect / grandTotal) * 100).toFixed(1) : '0.0';
+  const overallAccuracy = formatPct(totalCorrect, grandTotal);
```
[the remark this answers is not in the document]
Comment (jefho-rh) [reply packages/automl/frontend/src/app/components/run-results/AutomlModelDetailsModal/tabs/ConfusionMatrixTab.tsx:174]: The AutoML backend always returns square confusion matrices where row keys and column keys are identical. `ConfusionMatrixData` is consumed as a symmetric `Record<string, Record<string, number>>` A matrix like `{ A: { A: 5, B: 5 } }` with a missing `B` row key does not occur in practice.

File: packages/automl/frontend/src/app/components/run-results/AutomlModelDetailsModal/tabs/ConfusionMatrixTab.tsx [hunk 3]
```diff
@@ -172,146 +162,151 @@ const ConfusionMatrixTab: React.FC<TabContentProps> = ({
...
       </div>
+      <Flex
+        alignItems={{ default: 'alignItemsCenter' }}
+        spaceItems={{ default: 'spaceItemsSm' }}
+        className="pf-v6-u-mt-md pf-v6-u-font-size-sm pf-v6-u-color-200"
+        data-testid="confusion-matrix-legend"
+      >
+        <span className="automl-confusion-matrix__legend-dot" />
+        <span>Correct prediction</span>
+      </Flex>
```
[the remark this answers is not in the document]
Comment (jefho-rh) [reply packages/automl/frontend/src/app/components/run-results/AutomlModelDetailsModal/tabs/ConfusionMatrixTab.tsx:309]: The project convention (per `css-patternfly.md`) explicitly uses plain SCSS imports with BEM naming — not CSS modules. Every package in the repo follows this pattern: co-located `.scss` files imported as `import './Component.scss'` with package-specific prefixes (`automl-`). The CodeRabbit path instruction conflicts with the actual project standards.

**Review by GAUNSD — CHANGES_REQUESTED** [review]
## AI code review 🤖  🔍
`🤖: Claude Opus 4.6 (1M context)`

Solid feature work — the confusion matrix redesign is clean and well-tested. The backtest multi-select introduces a `useRef`-based persistence pattern that is cleverly done for tab switching, but has a stale-state gap on model switches and a latent nullish-coalescing bug. Neither is likely to crash in normal usage, but both would be good to fix before merging.

## Summary

| # | Severity | File | Finding |
|---|----------|------|---------|
| 1 | 🟡 Medium | `BacktestWindowChart.tsx` | Nullish coalescing on empty array — `[]` passes through `??` |
| 2 | 🟡 Medium | `AutomlModelDetailsModal.tsx` | `backtestMetricsRef` not cleared on model switch — stale metric keys |
| 3 | 🟡 Medium | `BacktestWindowChart.tsx` | Hardcoded `CHART_W_GRID = 440` can overflow narrow viewports |
| 4 | 🔵 Low | `tabConfig.ts` | `TabContentProps` widened with backtest-specific props |
| 5 | 🔵 Low | `BacktestWindowChart.tsx` | Removed `data-testid="backtest-window-chart"` from `BacktestCurveChart` |

## Details

<details><summary>Item 1: 🟡 Nullish coalescing on empty array doesn't fallback to `evalMetric`</summary>

### 🟡 Nullish coalescing on empty array doesn't fallback to `evalMetric`

**File:** `BacktestWindowChart.tsx`, line ~485 (new code)

```tsx
const [selectedMetrics, setSelectedMetrics] = React.useState<string[]>(
  initialSelectedMetrics ?? [evalMetric],
);
```

The `??` operator only checks for `null`/`undefined`, not falsy values. An empty array `[]` is truthy, so `[] ?? [evalMetric]` evaluates to `[]` — skipping the intended
fallback.

If `initialSelectedMetrics` is ever `[]`, the component initializes with zero selected metrics. `isSingleMetric` becomes `false` (length 0 ≠ 1), the multi-metric grid renders
with no items, and the toggle shows "0 metrics".

**Currently unreachable** because `updateMetrics` enforces a minimum of one element and `backtestMetricsRef` starts as `undefined`. But if any future caller passes `[]`, this
silently breaks.

**Fix:**
```tsx
const [selectedMetrics, setSelectedMetrics] = React.useState<string[]>(
  initialSelectedMetrics?.length ? initialSelectedMetrics : [evalMetric],
);
```
</details>

<details><summary>Item 2: 🟡 `backtestMetricsRef` not cleared on model switch — stale metric keys persist</summary>

### 🟡 `backtestMetricsRef` not cleared on model switch — stale metric keys persist

**File:** `AutomlModelDetailsModal.tsx`, line ~259 (new code)

```tsx
const backtestMetricsRef = React.useRef<string[]>();
```

The ref persists selected metric keys across the modal's lifetime. When the user switches to a different model via the model dropdown:

1. `setSelectedModelName(name)` triggers a re-render
2. `useModelEvaluationArtifactsQuery` fetches new `backTesting` data for the new model
3. `tabContentProps.backtestSelectedMetrics` still holds the old model's metric keys
4. If `BacktestWindowChart` stays mounted (same tab active), its `selectedMetrics` state retains the old keys

If the two models happen to have different metric sets (uncommon in the same AutoML run, but not impossible), the chart tries to render metric keys that don't exist in the new
model's data — producing empty/broken chart panels with stale toggle labels.

**Fix:** Clear the ref when the model changes:
```tsx
React.useEffect(() => {
  backtestMetricsRef.current = undefined;
}, [selectedModelName]);
```

Or reset `selectedMetrics` inside `BacktestWindowChart` when `perWindowMetrics` changes (e.g. via a key prop tied to the model name).
</details>

<details><summary>Item 3: 🟡 Hardcoded `CHART_W_GRID = 440` can overflow narrow viewports</summary>

### 🟡 Hardcoded `CHART_W_GRID = 440` can overflow narrow viewports

**File:** `BacktestWindowChart.tsx`, line ~30

```tsx
const CHART_W_GRID = 440;
```

The multi-metric grid renders two `BacktestMetricChart` components side-by-side in `<GridItem span={6}>`. Each chart has a fixed SVG width of 440px. The PF `Grid` column width
depends on the parent container, which is the modal body minus the sidebar. If the modal content area is narrower than ~880px (e.g. on a 1024px viewport, or a resized browser
window), the SVG overflows its grid cell.

The existing `CHART_W = 900` has the same limitation for the single-metric view, but at least it matches the modal's `variant="large"` width. Adding a second fixed width for a
2-column layout doubles the responsive breakage risk.

Not a code-freeze blocker since the existing pattern has the same issue, but worth noting for a follow-up since the multi-metric grid is a new interaction path that users will
discover.
</details>

<details><summary>Item 4: 🔵 `TabContentProps` widened with backtest-specific props</summary>

### 🔵 `TabContentProps` widened with backtest-specific props

**File:** `tabConfig.ts`, lines ~36-37

```tsx
export type TabContentProps = {
  ...
  backtestSelectedMetrics?: string[];
  onBacktestMetricsChange?: (metrics: string[]) => void;
};
```

These two props only matter for `BacktestingTab`. Every other tab component (`ModelInformationTab`, `FeatureSummaryTab`, `ConfusionMatrixTab`, etc.) receives them as unused
optional props. This isn't a bug, but it widens a shared interface with single-consumer concerns.

As more tabs get UI state persistence (e.g. if the confusion matrix view selector also needs to persist), this type will accumulate props that are irrelevant to most consumers.
Consider a narrower pattern (intersection type at the call site, or a generic persistence mechanism) in a follow-up.
</details>

<details><summary>Item 5: 🔵 Removed `data-testid` from `BacktestCurveChart` usage</summary>

### 🔵 Removed `data-testid` from `BacktestCurveChart` usage

**File:** `BacktestWindowChart.tsx` (old code removed)

The old code passed `data-testid="backtest-window-chart"` as a prop to `<BacktestCurveChart>`. This prop was removed in the refactor to `BacktestMetricChart`. Since
`BacktestCurveChart` is a custom component that constructs a PF Victory `<Chart>` and doesn't spread unknown props, the `data-testid` likely never reached the DOM — so this is
probably a no-op removal. But worth confirming no Cypress selectors rely on it.
</details>

**Comment by jefho-rh** [comment]
Thanks for the review @GAUNSD! 

Addressed items 1, 2, 3, and 5. Item 4 deferred.

### Fixed

| # | Finding | Fix |
|---|---------|-----|
| 1 | `??` on empty array doesn't fallback to `evalMetric` | Changed to `initialSelectedMetrics?.length ? initialSelectedMetrics : [evalMetric]` |
| 2 | `backtestMetricsRef` not cleared on model switch | Added `useEffect` to reset ref when `selectedModelName` changes |
| 3 | Hardcoded `CHART_W_GRID = 440` overflows narrow viewports | Reduced to 400 (adds ~40-50px margin in grid cells) and added `overflow: hidden` on the card as a safety net |
| 5 | Removed `data-testid="backtest-window-chart"` | Restored on `BacktestCurveChart` inside `BacktestMetricChart` — Cypress page object (`resultsPage.ts:179`) depends on it |

### Deferred

| # | Finding | Reason |
|---|---------|--------|
| 4 | `TabContentProps` widened | Will revisit in a followup PR|

**Comment by GAUNSD** [comment]
Thanks for all the fixes @jefho-rh!
/approve
/lgtm
